### ISSUE TITLE
[arethetypeswrong] Plugin types are wrong

### ISSUE THREAD

**Issue by bradzacher** [eslint-stylistic/eslint-stylistic#481]
### Validations

- [X] Follow our [Code of Conduct](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://eslint.style/contribute/guide).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### Describe the bug

https://arethetypeswrong.github.io/?p=%40stylistic%2Feslint-plugin%402.4.0

> [Incorrect default export](https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseExportDefault.md)
> 
> The resolved types use export default where the JavaScript file appears to use `module.exports =`. This will cause TypeScript under the `node16` module mode to think an extra `.default` property access is required, but that will likely fail at runtime. These types should use `export =` instead of export default.

### Reproduction

[TS playground](https://www.typescriptlang.org/play/?target=99&moduleResolution=99&module=100#code/JYWwDg9gTgLgBGANgVwObAHZwGZQiOAcgAEBnGAT0WHOAGMB6AU1OoxgFok1NCBuAFADu6DADooyRCwFA)

**Comment by bradzacher** [comment]
Yeah sadly TS doesn't allow you to have both an `export =` and an ESM export -- even if the other ESM exports are just types.

One option is to use a `namespace` to merge the types onto the `export =`'d name, eg:

```ts
import { Linter, ESLint, Rule } from 'eslint';
import { UnprefixedRuleOptions as UnprefixedRuleOptions_ } from './rule-options.js';
import { RuleOptions as RuleOptions_ } from './rule-options.js';
import '@stylistic/eslint-plugin-js';
import '@stylistic/eslint-plugin-jsx';
import '@stylistic/eslint-plugin-ts';
import '@stylistic/eslint-plugin-plus';

declare namespace plugin {
    export type Rules = { ... };
    export interface StylisticCustomizeOptions<Flat extends boolean = true> { ... }
    export type UnprefixedRuleOptions = UnprefixedRuleOptions_;
    export type RuleOptions = UnprefixedRuleOptions_;
}
declare function customize(options: plugin.StylisticCustomizeOptions<false>): Linter.BaseConfig;
declare function customize(options?: plugin.StylisticCustomizeOptions<true>): Linter.Config;

declare const _configs: { ... };
declare const configs: ESLint.Plugin["configs"] & typeof _configs;

declare const plugin: {
    rules: plugin.Rules;
    configs: typeof configs;
};

export = plugin;
```

This will allow people to access the types via `type T = plugin.Rules;`, etc.
But this does mean that people can't import the types directly like `import { type Rules } from  '@stylistic/eslint-plugin'`.

---

I believe that explicitly setting `type: "commonjs"` would also fix it without any changes to the types.
For example for `typescript-eslint` that's what we do and it works as expected.

---

> I wanted to push ESLint ecosystem towards ESM since now we have the flat config. I don't really want to go backward to use CJS conversions

I definitely understand this! But until your tool fully drop support for eslintrc configs you're stuck supporting CJS 😢.

If you want to dual-emit CJS+ESM then there are tricks you can do to support types for that situation but things do get a bit complicated IIRC.

**Comment by antfu** [comment]
> I believe that explicitly setting type: "commonjs"

Do you mean by doing it in the package.json? It doesn't seems to solve it from my local reproduction

**Comment by bradzacher** [comment]
I would compare things with our `typescript-eslint` package which is a CJS packaged authored in `.ts` files in ESM and built to `.js` files in CJS.

https://arethetypeswrong.github.io/?p=typescript-eslint%408.1.0

I think the main difference is the use of `type: commonjs`? Perhaps there's something else I missed.

**Comment by antfu** [comment]
As this plugin is moved to ESM-only, I consider the runtime/type inconsistency is resolved. Thanks for bringing this up
